tsdb/head.go

```diff
@@ -1137,6 +1141,23 @@ func (a *headAppender) AddFast(ref uint64, t int64, v float64) error {
 	if s == nil {
 		return errors.Wrap(storage.ErrNotFound, "unknown series")
 	}
+
+	// Without string interning / using symbol table / using siginficantly
+	// more memory, determine whether the metadata has changed for this entry
+	// using a hash of the metadata.
+	a.digest.Reset()
```

Comment:
This sounds like it's going to be quite expensive for what is the most important hot path in all of Prometheus. We've also had quite a few problems in the past with hashes and collisions. If we're going to compare, let's do it literally.

Or leave it up to the scrape, so using Add implies new metadata and AddFast is always fast.

Reply:
That's fair, are you open to changing the appender interface to the following:

```go
Add(l labels.Labels, m Metadata, t int64, v float64) (uint64, error)
AddFast(ref uint64, int64, v float64) error
AndFastAndUpdate(ref uint64, m Metadata, t int64, v float64) error
```

Then we can call `AndFastAndUpdate` if and only if the metadata has been updated (which can be done from the scrape loop by doing literal comparison).

Reply:
Caught up about this, seems like we can just revert to calling `Add(...)` if the metadata has changed between scrape iterations and remove the `AddFastAndUpdate(...)` concept.